src/Microsoft.IdentityModel.Tokens/LockingLRUCache.cs

```diff
@@ -0,0 +1,283 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.CryptoProviderCacheOptions
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.Logging;
+
+namespace Microsoft.IdentityModel.Tokens
+{
+    internal class LockingLRUCache<TKey, TValue> : ILRUCache<TKey, TValue>, IDisposable
+    {
+        private int _capacity;
+        private int _count = 0;
+        private Dictionary<TKey, LinkedListNode<CacheItem<TKey, TValue>>> _map;
+        private LinkedList<CacheItem<TKey, TValue>> _doubleLinkedList = new LinkedList<CacheItem<TKey, TValue>>();
+        // Used to ensure that the cache is thread-safe.
+        private object _cacheLock = new object();
+        private readonly BlockingCollection<Action> _cleanupEventQueue = new BlockingCollection<Action>();
+        private bool _disposed = false;
+
+        internal LockingLRUCache(int capacity, IEqualityComparer<TKey> comparer = null)
+        {
+            _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _map = new Dictionary<TKey, LinkedListNode<CacheItem<TKey, TValue>>>(comparer ?? EqualityComparer<TKey>.Default);
+
+            var thread = new Thread(new ThreadStart(OnStart));
+            thread.IsBackground = true;
+            thread.Start();
+
+            Task timerTask = RemoveExpiredValuesPeriodically(TimeSpan.FromMinutes(5));
+        }
+
+        private void OnStart()
+        {
+            while (true)
+            {
+                _cleanupEventQueue.Take().Invoke();
+            }
+        }
+
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        public int RemoveExpiredValues()
+        {
+            int numItemsRemoved = 0;
+            var node = _doubleLinkedList.First;
```

Comment:
I believe this is a race condition. If this gets called directly by a user and from the periodic function, you don't have a guarantee that first will be the same after a run of the loop below.

Reply:
I do think that could be a race condition...

I've made the method internal for now, as it was only public so that we could have an interface to easily switch between the two caching models. I believe this removes the potential for a race condition, but if we decide we need to expose the method publicly this will need to be looked at again. https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1575/commits/45c168bd46aa16f03a5e5f83c8b9b2b8cef9b070